### PR TITLE
Expand logging configuration support in PennyLane, and also target Catalyst

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** To better unify PennyLane's logging support, updates to control PennyLane's logging behaviour enabling use with Catalyst allow a unified way to examine execution pipelines throughout the Pythonic layer. This PR expands the configuration of logging, allowing direct editing by a user, as well as adds new module-level log-levels for Catalyst, and expands jax to also enable logging with jaxlib. The above will be useful to define interactions across components in the ecosystem, and aid in building a model of explicit interactions across the packages.

**Description of the Change:** As above.

**Benefits:** Expands the logging framework to be more configurable, and extends upcoming support for Catalyst.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by trbromley — COMMENTED** [review]
Thanks @mlxd! Please could you tag the corresponding Shortcut story?

**Comment by mlxd** [comment]
[sc-61591]

**Review by dime10 — COMMENTED** [review]
The changes seem okay to me, but are they really about targeting Catalyst as the title suggests? (I could only find a few lines related to Catalyst)

File: pennylane/logging/utils.py
Comment (mudit2812) [inline pennylane/logging/utils.py:23]:
```suggestion
    Predicate that validates if argument ``f`` is a local function belonging to module ``mod_name``.
```

File: pennylane/logging/utils.py
Comment (mudit2812) [inline pennylane/logging/utils.py:32]:
```suggestion
    Modifies the module ``mod_name`` to add logging implicitly to all free-functions.
```

**Review by mudit2812 — COMMENTED** [review]
Love the decorators! Just curious on your thoughts about adding a decorator which can be applied to a class which would automatically decorate the class's methods using a similar approach as the `_add_logging_all` function you added? I guess right now we want to avoid decorating a whole class so that there is more fine-grained control over which functions actually trigger a logging call.

Overall looking great! Just waiting on tests before approval.

**Comment by trbromley** [comment]
Thanks @mlxd! It might be worth making sure the [logging quickstart](https://docs.pennylane.ai/en/stable/introduction/logging.html) and [logging dev guide](https://docs.pennylane.ai/en/stable/development/guide/logging.html) are up dated as needed.

**Comment by mlxd** [comment]
> The changes seem okay to me, but are they really about targeting Catalyst as the title suggests? (I could only find a few lines related to Catalyst)

Title "to" changed to "and", which should be more appropriate now.. Most of the infra changes in here involved a PR merged onto this one, which adds the decorators and ensures they work too in PennyLane.

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py:73]: Good catch.

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py:114]: Given that this function isn't needed in most cases, I'd rather keep it at this level to avoid the inherent module level lookup, since those can slow down for scaled out workloads at startup.

File: pennylane/logging/formatters/formatter.py
Comment (mlxd) [reply pennylane/logging/formatters/formatter.py:118]: Good call --- I'll update accordingly.

File: pennylane/logging/utils.py
Comment (mlxd) [reply pennylane/logging/utils.py:30]: More of a quick 'n' dirty debugging option for free functions --- we can add it in at the bottom of a module as needed, which is why we can keep it private --- I'd rather users aren't aware of this.

File: pennylane/logging/__init__.py
Comment (mlxd) [reply pennylane/logging/__init__.py:20]: This is for debug-purposes only with Cat/LQ. I need it added here to explore its use with the other packages.

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py:85]: Done

File: pennylane/logging/formatters/formatter.py
Comment (mlxd) [reply pennylane/logging/formatters/formatter.py:118]: Done.

**Comment by mlxd** [comment]
> Thanks @mlxd! It might be worth making sure the [logging quickstart](https://docs.pennylane.ai/en/stable/introduction/logging.html) and [logging dev guide](https://docs.pennylane.ai/en/stable/development/guide/logging.html) are up dated as needed.

Good call, and done.

File: doc/development/guide/logging.rst
Comment (mudit2812) [inline doc/development/guide/logging.rst:34]: These points as well as the code blocks within need to be indented to render correctly.

File: doc/development/guide/logging.rst
Comment (mudit2812) [inline doc/development/guide/logging.rst:38]:
```suggestion
   # debug_logger can be used to decorate any method or free function
```

File: doc/development/guide/logging.rst
Comment (mudit2812) [inline doc/development/guide/logging.rst:97]:
```suggestion
For ease-of-development, the function :func:`pennylane.logging.edit_system_config` opens an editor (if the ``EDITOR`` environment variable is set), or a viewer of the existing file configuration, which can be used to modify the existing options.
```

**Review by mudit2812 — APPROVED** [review]
Minor suggestions about fixing how docs are rendering, but otherwise LGTM!

File: doc/development/guide/logging.rst
Comment (mlxd) [reply doc/development/guide/logging.rst]: Good call

File: tests/logging/test_logging_autograd.py
Comment (mlxd) [inline tests/logging/test_logging_autograd.py:149]:
```suggestion
```

File: tests/logging/test_logging_autograd.py
Comment (mlxd) [inline tests/logging/test_logging_autograd.py:152]:
```suggestion
```

File: tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py
Comment (mlxd) [inline tests/devices/qutrit_mixed/test_qutrit_mixed_tracking.py:228]: Note: we do not want devices to add logging tests explicitly, due to the potential for introduced slowdowns in the event of failure. This is mitigated as best as possible in the stack, but logging show be validated independently of devices, hence the removal.

**Comment by mlxd** [comment]
FYI, the CodeCov complaints are due to the newly added untestable source. We ignore these as pytest-runs are negatively impacted in CI runtime if these are hit.

File: pennylane/logging/configuration.py
Comment (trbromley) [inline pennylane/logging/configuration.py:92]: So to make sure I understand, the default log file won't be in the user's current working directory?

File: pennylane/logging/configuration.py
Comment (trbromley) [inline pennylane/logging/configuration.py:134]:
```suggestion
    Setting the ``EDITOR`` environment variable will override ``xdg-open/open`` on
    Linux and MacOS, and allows use of ``wait_on_close`` for editor close before
    continuing execution.
```

File: pennylane/logging/configuration.py
Comment (trbromley) [inline pennylane/logging/configuration.py:139]:
```suggestion
        instead open this file manually with the ``config_path()`` provided path.
```

File: pennylane/logging/configuration.py
Comment (trbromley) [inline pennylane/logging/configuration.py:141]: Is this environment variable normally set up on systems without users having to do anything? It didn't work for me.

**Review by trbromley — COMMENTED** [review]
Thanks @mlxd

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py:92]: Nope, nor has it ever been. It is a TOML file we ship with PennyLane, that can be overriden by a suer, or edit the in-built one using the newly provided commands.

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py:141]: Most linux systems set this already, or most dev users set this themselves. If not set, the backup is to open the file with the browser.

File: pennylane/logging/configuration.py
Comment (trbromley) [reply pennylane/logging/configuration.py:92]: That makes sense, I remember now. Do you think it's worth adding a reminder that `config_path` can be used to find the current location

File: pennylane/logging/configuration.py
Comment (trbromley) [reply pennylane/logging/configuration.py:141]: The only data point I have is that it didn't work for me :laughing:

So are we sure this will work out of the box for most users? If not, is this function enough of a utility to be worth it?

Comment (mlxd) [reply pennylane/logging/configuration.py:141]: Seems like parenthesis were needed to properly scope the if-else and walrus operator. Should be fixed now.

File: pennylane/logging/configuration.py
Comment (mlxd) [reply pennylane/logging/configuration.py]: Added a comment and a ref to the config_path function to query the absolute path. Will mark as resolved.

File: pennylane/logging/configuration.py
Comment (trbromley) [reply pennylane/logging/configuration.py:141]: Can confirm, it works for me now :star_struck:

**Review by trbromley — APPROVED** [review]

**Pull request merged**
